pkg/infrastructure/aws/clusterapi/aws.go

```diff
@@ -24,3 +34,112 @@ func (*Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionInp
 	}
 	return nil
 }
+
+// InfraReady creates private hosted zone and DNS records.
+func (*Provider) InfraReady(ctx context.Context, in clusterapi.InfraReadyInput) error {
+	awsCluster := &capa.AWSCluster{}
+	key := k8sClient.ObjectKey{
+		Name:      in.InfraID,
+		Namespace: capiutils.Namespace,
+	}
+	if err := in.Client.Get(ctx, key, awsCluster); err != nil {
+		return fmt.Errorf("failed to get AWSCluster: %w", err)
+	}
+
+	awsSession, err := in.InstallConfig.AWS.Session(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get aws session: %w", err)
+	}
+
+	subnetIDs := make([]string, 0, len(awsCluster.Spec.NetworkSpec.Subnets))
+	for _, s := range awsCluster.Spec.NetworkSpec.Subnets {
+		if s.IsPublic {
+			subnetIDs = append(subnetIDs, s.ResourceID)
+		}
+	}
+
+	vpcID, err := getVPCFromSubnets(ctx, awsSession, awsCluster.Spec.Region, subnetIDs)
+	if err != nil {
+		return err
+	}
+
+	tags := map[string]string{
+		fmt.Sprintf("kubernetes.io/cluster/%s", in.InfraID): "owned",
+	}
+	for k, v := range awsCluster.Spec.AdditionalTags {
+		tags[k] = v
+	}
+
+	client := awsconfig.NewClient(awsSession)
+
+	phzID := in.InstallConfig.Config.AWS.HostedZone
+	if len(phzID) == 0 {
+		logrus.Infoln("Creating private Hosted Zone")
+		res, err := client.CreateHostedZone(ctx, &awsconfig.HostedZoneInput{
+			InfraID:  in.InfraID,
+			VpcID:    vpcID,
+			Region:   awsCluster.Spec.Region,
+			Name:     in.InstallConfig.Config.ClusterDomain(),
+			Role:     in.InstallConfig.Config.AWS.HostedZoneRole,
+			UserTags: tags,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create private hosted zone: %w", err)
+		}
+		phzID = aws.StringValue(res.Id)
+	}
+
+	logrus.Infoln("Creating Route53 records for control plane load balancer")
+	apiHost := awsCluster.Status.Network.SecondaryAPIServerELB.DNSName
+	logrus.Debugf("apiHost = %s", apiHost)
+	apiIntHost := awsCluster.Spec.ControlPlaneEndpoint.Host
+	logrus.Debugf("apiIntHost = %s", apiIntHost)
```

Comment:
do we need to check if those records exists?

For the secondary LB, the PR https://github.com/openshift/installer/pull/8149 is adding as optional, when publish==External. Do we need the same check here? 

What is the value of DNS RR `api.` when publish==Internal?

Reply:
If publish is not "External" then `awsCluster.Status.Network.SecondaryAPIServerELB.DNSName` will be an empty string (because the second LB won't exist). No record will be created for that, as `CreateOrUpdateRecord` will skip the creation of public record for public zone.